File: java/src/test/java/kanzi/test/TestHash.java
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.test;

import java.io.File;
import java.io.FileInputStream;
import kanzi.util.hash.MurMurHash3;
import kanzi.util.hash.SipHash_2_4;
import kanzi.util.hash.XXHash32;
import kanzi.util.hash.XXHash64;


public class TestHash
{
  public static void main(String[] args)
  {    
      try
      {
         String fileName = (args.length > 0) ? args[0] : "c:\\temp\\rt.jar";
         int iter = 500;
         System.out.println("Processing "+fileName);
         System.out.println(iter+" iterations");

         {
            System.out.println("XXHash32 speed test");
            File input = new File(fileName);
            FileInputStream fis = new FileInputStream(input);
            byte[] array = new byte[16384];
            XXHash32 hash = new XXHash32(0);
            int len;
            long size = 0;
            int res = 0;
            long sum = 0;

            while ((len = fis.read(array, 0, array.length)) > 0)
            { 
               long before = System.nanoTime();
               
               for (int i=0; i<iter; i++)
               {
                  hash.setSeed(0);
                  res += hash.hash(array, 0, len);
               }
               
               long after = System.nanoTime();
               sum += (after - before);
               size += (len * iter);
            }
           
            fis.close();
            System.out.println("XXHash32 res="+Integer.toHexString(res));
            System.out.println("Elapsed [ms]: " +sum/1000000L);
            System.out.println("Throughput [MB/s]: " +(size/1024*1000/1024)/(sum/1000000L));
         }
         
         System.out.println();
         
         {
            System.out.println("XXHash64 speed test");
            File input = new File(fileName);
            FileInputStream fis = new FileInputStream(input);
            byte[] array = new byte[16384];
            XXHash64 hash = new XXHash64(0);
            int len;
            long size = 0;
            long res = 0;
            long sum = 0;

            while ((len = fis.read(array, 0, array.length)) > 0)
            { 
               long before = System.nanoTime();
               
               for (int i=0; i<iter; i++)
               {
                  hash.setSeed(0);
                  res += hash.hash(array, 0, len);
               }
               
               long after = System.nanoTime();
               sum += (after - before);
               size += (len * iter);
            }
           
            fis.close();
            System.out.println("XXHash64 res="+Long.toHexString(res));
            System.out.println("Elapsed [ms]: " +sum/1000000L);
            System.out.println("Throughput [MB/s]: " +(size/1024*1000/1024)/(sum/1000000L));
         }
         
         System.out.println();
         
         {
            System.out.println("MurmurHash3 speed test");
            File input = new File(fileName);
            FileInputStream fis = new FileInputStream(input);
            byte[] array = new byte[16384];
            MurMurHash3 hash = new MurMurHash3(0);
            int len;
            long size = 0;
            int res = 0;
            long sum = 0;

            while ((len = fis.read(array, 0, array.length)) > 0)
            { 
               long before = System.nanoTime();

               for (int i=0; i<iter; i++)
               {
                  hash.setSeed(0);
                  res += hash.hash(array, 0, len);     
               }               

               long after = System.nanoTime();
               sum += (after - before);
               size += (len * iter);
            }            
            
            fis.close();
            System.out.println("MurmurHash res="+Integer.toHexString(res));
            System.out.println("Elapsed [ms]: " +sum/1000000L);
            System.out.println("Throughput [MB/s]: " +(size/1024*1000/1024)/(sum/1000000L));
         }

         
         System.out.println();
         
         {
            System.out.println("SipHash_2_4 speed test");
            File input = new File(fileName);
            FileInputStream fis = new FileInputStream(input);
            byte[] array = new byte[16384];
            SipHash_2_4 hash = new SipHash_2_4();
            int len;
            long size = 0;
            long res = 0;
            long sum = 0;

            while ((len = fis.read(array, 0, array.length)) > 0)
            { 
               long before = System.nanoTime();
   
               for (int i=0; i<iter; i++)
               {
                  hash.setSeed(0, 0);
                  res += hash.hash(array, 0, len);     
               }
               
               long after = System.nanoTime();
               sum += (after - before);
               size += (len * iter);
            }            
            
            fis.close();
            System.out.println("SipHash_2_4 res="+Long.toHexString(res));
            System.out.println("Elapsed [ms]: " +sum/1000000L);
            System.out.println("Throughput [MB/s]: " +(size/1024*1000/1024)/(sum/1000000L));
         }
      }
      catch (Exception e)
      {
         e.printStackTrace();
      }
  }  
}

File: test/TestHash.go
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"fmt"
	"github.com/flanglet/kanzi/util/hash"
	"os"
	"time"
)

func main() {
	var filename = flag.String("input", "c:\\temp\\rt.jar", "name of the input file")
	flag.Parse()
	iter := 500
	fmt.Printf("Processing %v\n", *filename)
	fmt.Printf("%v iterations\n", iter)

	{
		fmt.Printf("XXHash32 speed test\n")
		file, err := os.Open(*filename)

		if err != nil {
			fmt.Printf("Cannot open %s", *filename)

			return
		}

		defer file.Close()
		buffer := make([]byte, 16384)
		hash, err := hash.NewXXHash32(0)

		if err != nil {
			fmt.Printf("Failed to create hash: %v\n", err)
			return
		}

		length, err := file.Read(buffer)
		size := int64(0)
		res := uint32(0)
		sum := int64(0)

		for length > 0 {
			if err != nil {
				fmt.Printf("Failed to read the next chunk of input file '%v': %v\n", *filename, err)
				return
			}

			before := time.Now()

			for i := 0; i < iter; i++ {
				hash.SetSeed(0)
				res += hash.Hash(buffer[0:length])
			}

			after := time.Now()
			sum += after.Sub(before).Nanoseconds()
			size += int64(length * iter)
			length, err = file.Read(buffer)
		}

		sum /= 1000000 // convert to ms
		fmt.Printf("XXHash32 res=%x\n", res)
		fmt.Printf("Elapsed [ms]: %v\n", sum)
		fmt.Printf("Throughput [MB/s]: %v\n", (size/1024*1000/1024)/sum)
	}

	fmt.Printf("\n")

	{
		fmt.Printf("XXHash64 speed test\n")
		file, err := os.Open(*filename)

		if err != nil {
			fmt.Printf("Cannot open %s", *filename)

			return
		}

		defer file.Close()
		buffer := make([]byte, 16384)
		hash, err := hash.NewXXHash64(0)

		if err != nil {
			fmt.Printf("Failed to create hash: %v\n", err)
			return
		}

		length, err := file.Read(buffer)
		size := int64(0)
		res := uint64(0)
		sum := int64(0)

		for length > 0 {
			if err != nil {
				fmt.Printf("Failed to read the next chunk of input file '%v': %v\n", *filename, err)
				return
			}

			before := time.Now()

			for i := 0; i < iter; i++ {
				hash.SetSeed(0)
				res += hash.Hash(buffer[0:length])
			}

			after := time.Now()
			sum += after.Sub(before).Nanoseconds()
			size += int64(length * iter)
			length, err = file.Read(buffer)
		}

		sum /= 1000000 // convert to ms
		fmt.Printf("XXHash64 res=%x\n", res)
		fmt.Printf("Elapsed [ms]: %v\n", sum)
		fmt.Printf("Throughput [MB/s]: %v\n", (size/1024*1000/1024)/sum)
	}

	fmt.Printf("\n")

	{
		fmt.Printf("MurmurHash3 speed test\n")
		file, err := os.Open(*filename)

		if err != nil {
			fmt.Printf("Cannot open %s", *filename)

			return
		}

		defer file.Close()
		buffer := make([]byte, 16384)
		hash, err := hash.NewMurMurHash3(uint32(0))

		if err != nil {
			fmt.Printf("Failed to create hash: %v\n", err)
			return
		}

		length, err := file.Read(buffer)
		size := int64(0)
		res := uint32(0)
		sum := int64(0)

		for length > 0 {
			if err != nil {
				fmt.Printf("Failed to read the next chunk of input file '%v': %v\n", *filename, err)
				return
			}

			before := time.Now()

			for i := 0; i < iter; i++ {
				hash.SetSeed(0)
				res += hash.Hash(buffer[0:length])
			}

			after := time.Now()
			sum += after.Sub(before).Nanoseconds()
			size += int64(length * iter)
			length, err = file.Read(buffer)
		}

		sum /= 1000000 // convert to ms
		fmt.Printf("MurmurHash3 res=%x\n", res)
		fmt.Printf("Elapsed [ms]: %v\n", sum)
		fmt.Printf("Throughput [MB/s]: %v\n", (size/1024*1000/1024)/sum)
	}

	fmt.Printf("\n")

	{
		fmt.Printf("SipHash_2_4 speed test\n")
		file, err := os.Open(*filename)

		if err != nil {
			fmt.Printf("Cannot open %s", *filename)

			return
		}

		defer file.Close()
		buffer := make([]byte, 16384)
		hash, err := hash.NewSipHash()

		if err != nil {
			fmt.Printf("Failed to create hash: %v\n", err)
			return
		}

		length, err := file.Read(buffer)
		size := int64(0)
		res := uint64(0)
		sum := int64(0)

		for length > 0 {
			if err != nil {
				fmt.Printf("Failed to read the next chunk of input file '%v': %v\n", *filename, err)
				return
			}

			before := time.Now()

			for i := 0; i < iter; i++ {
				hash.SetSeedFromLongs(0, 0)
				res += hash.Hash(buffer[0:length])
			}

			after := time.Now()
			sum += after.Sub(before).Nanoseconds()
			size += int64(length * iter)
			length, err = file.Read(buffer)
		}

		sum /= 1000000 // convert to ms
		fmt.Printf("SipHash_2_4 res=%x\n", res)
		fmt.Printf("Elapsed [ms]: %v\n", sum)
		fmt.Printf("Throughput [MB/s]: %v\n", (size/1024*1000/1024)/sum)
	}
}
